chainio/blockbeat.go

```diff
@@ -0,0 +1,154 @@
+package chainio
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/btcsuite/btclog"
+	"github.com/lightningnetwork/lnd/build"
+	"github.com/lightningnetwork/lnd/chainntnfs"
+	"github.com/lightningnetwork/lnd/fn"
+)
+
+// DefaultProcessBlockTimeout is the timeout value used when waiting for one
+// consumer to finish processing the new block epoch.
+var DefaultProcessBlockTimeout = 30 * time.Second
+
+// Beat implements the Blockbeat interface. It contains the block epoch and a
+// buffer error chan.
+//
+// TODO(yy): extend this to check for confirmation status - which serves as the
+// single source of truth, to avoid the potential race between receiving blocks
+// and `GetTransactionDetails/RegisterSpendNtfn/RegisterConfirmationsNtfn`.
+type Beat struct {
+	// epoch is the current block epoch the blockbeat is aware of.
+	epoch chainntnfs.BlockEpoch
+
+	// errChan is a buffered chan that receives an error returned from
+	// processing this block.
+	errChan chan error
+
+	// log is the customized logger for the blockbeat which prints the
+	// block height.
+	log btclog.Logger
+}
+
+// Compile-time check to ensure Beat satisfies the Blockbeat interface.
+var _ Blockbeat = (*Beat)(nil)
+
+// NewBeat creates a new beat with the specified block epoch and a buffered
+// error chan.
+func NewBeat(epoch chainntnfs.BlockEpoch) Beat {
+	b := Beat{
+		epoch:   epoch,
+		errChan: make(chan error, 1),
+	}
+
+	// Create a customized logger for the blockbeat.
+	logPrefix := fmt.Sprintf("Height[%6d]:", b.Height())
+	b.log = build.NewPrefixLog(logPrefix, clog)
+
+	return b
+}
+
+// NewBeatFromHeight creates a new beat with the specified height and a
+// buffered error chan. This method should only be used in test.
+func NewBeatFromHeight(height int32) Beat {
+	return NewBeat(chainntnfs.BlockEpoch{Height: height})
+}
+
+// Height returns the height of the block epoch.
+//
+// NOTE: Part of the Blockbeat interface.
+func (b Beat) Height() int32 {
+	return b.epoch.Height
+}
+
+// NOTE: Part of the Blockbeat interface.
+func (b Beat) NotifyBlockProcessed(err error, quitChan chan struct{}) {
+	fn.SendOrQuit(b.errChan, err, quitChan)
+}
+
+// DispatchSequential takes a list of consumers and notify them about the new
+// epoch sequentially.
+//
+// NOTE: Part of the Blockbeat interface.
+func (b Beat) DispatchSequential(consumers []Consumer) error {
```

Comment:
Are all these methods implemented on a non-pointer receiver on purpose?
I think for channels it's fine as they are pointers. But if we have other members, such as a wait group or locks, we would run into an issue if `Beat` was copied instead of being accessed as a pointer.

Reply:
The final iteration now no longer has this design.